### DISCUSSION TITLE
Closing Unused Buffers When Disabling Tabs

### DISCUSSION THREAD

**Discussion by eric-hemasystems** [orbitalquark/textadept#264] [Show and tell]
I was glad to see Textadept had the `ui.tabs = false` configuration. I never find tabs useful. If I want to re-visit a file I previous had open I generally use the quick open dialog as I can just type a few characters of the filename and hit enter vs trying to figure out if it's an already open tab and where it is in the tab list.

But just disabling the tabs left the buffers open. Some might find this useful as they want to use the open buffer dialog or something. But again I tend to prefer to just hit the quick open dialog so I would rather the new buffer replace the old buffer not add it to a stack.

Luckily Textadept is so customizable I was able to make this happen with:

```lua
function contains(table, evaluator)
  for _, item in ipairs(table) do
    if evaluator(item) then return true end
  end

  return false
end

function close_unused_buffers()
  for _, buffer in ipairs(_BUFFERS) do
    local is_use = contains(_VIEWS, function(view) return view.buffer == buffer end)
    if not is_use then buffer:close() end
  end
end

ui.tabs = false
events.connect(events.FILE_OPENED, close_unused_buffers)
events.connect(events.BUFFER_AFTER_SWITCH, close_unused_buffers)
```

Thought I would share it in case anybody else has the same desire to disable tabs and have new buffers replace the old buffer vs add the new buffer on top of the old buffer.

---

Side question, I'm new to Lua. Does it seriously not have a function to test if an item is in an list? Seems like something that should just be part of the stdlib. It was easy to implement but still....

**Comment by orbitalquark** [comment]
Thanks for sharing! It's funny you bring it up because for the longest time Textadept did not have tabs (I never needed them). When I finally got around to adding them, it naturally had to be configurable :)

You are correct that Lua does not have a `table.find()` or similar method.

**Reply by eric1234** [reply]
The above code has actually proven to be unreliable. I've been able to come up with steps that get to the core of the issue. With the stock configuration (no need for the code above) if you do the following:

1. Start with a single view (view 1)
2. Open a file (buffer 1)
3. Split the view vertically (`ctrl`-`alt`-`v`)
4. In the right view (view 2) open a different file
5. Now open another different file in the view 2.
6. Split the view 2 creating view 3
7. In view 3 open yet another different file

End result is you should have 3 views but 4 buffers open. Now we want to attempt to close the unused buffer. This should be located at `_BUFFERS[2]` so run the following in the Lua REPL:

```lua
(function() _BUFFERS[2]:close() end)()
```

(wrapped close in anon function so the return value won't impact anything). If everything is working then you *should* not see any change as we are just closing the unused buffer. But what actually happens is that view #3 also switches to the same buffers are view 2 (causing buffer 4 to now also be unused even though I didn't ask view 3 to change buffers).

Any advice on what I am missing is greatly appreciated.

**Comment by orbitalquark** [comment]
It's because of this line: https://github.com/orbitalquark/textadept/blob/f6ed0a601b10981c6f0bca0c61fc3155a5cb2ed0/src/textadept.c#L1022

If you change "focused_view" to "dummy_view", then the focused view's buffer does not change. I am not completely sure if there would be any undesirable side effects for making this change. I'll have to experiment. My initial test worked without issue though.

**Reply by eric-hemasystems** [reply]
Glad you were able to track it down. I am working around it for now by just saving the association between views and buffers before closing the buffer, then resetting to that same association after. Seems to work so no hurry on getting it fixed if you want to give it time to make sure there are no side effects.

**Comment by orbitalquark** [comment]
Hmm, after further investigation, the culprit is actually here: https://github.com/orbitalquark/textadept/blob/f6ed0a601b10981c6f0bca0c61fc3155a5cb2ed0/core/ui.lua#L403

Long ago it was the case that you couldn't call `buffer:close()` on a hidden buffer, so that function to restore the previous buffer always assumed the current view just closed the closed/deleted buffer. I don't really know the best way to handle this, so I've just committed an expected failing test case for my TODO list: https://github.com/orbitalquark/textadept/commit/4e032ed63e4cc89da1396c1523ec9c012650c90b. Fortunately you have a workaround for yourself, as this is not high on my priority list.
